pydomcfg/domzgr/zgr.py

```diff
@@ -0,0 +1,154 @@
+
+                 sigma =  0 at the shallower boundary
+                 sigma = -1 at the deeper boundary
+
+        Parameters
+        ----------
+        k: int
+          Model level index. Note that
+             *) T-points are at integer values (between 1 and jpk)
+             *) W-points are at integer values - 1/2 (between 0.5 and jpk-0.5)
+        grd: str
+          If we are dealing with "T" or "W" model levels.
+
+        Returns
+        -------
+        float
+        """
+
+        kindx = float(k + 1)  # to deal with python convention
+        if grd == "W":
+            kindx -= 0.5
+
+        ps = -(kindx - 0.5) / float(self._jpk - 1)
+        return ps
+
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def compute_z3(
+        su: float,
+        ss1: float,
+        a1: float,
+        a2: float,
+        a3: float,
+        ss2: float = 0.0,
+        a4: float = 0.0,
+    ) -> float:
+        """
+        Generalised function providing the analytical
+        transformation from computational space to
+        physical space. It takes advantage of the fact that
+        z-coordinates can be considered a special case of
+        s-coordinate.
+
+        Note that z is downward positive.
+
+        Parameters
+        ----------
+        su: float
+            uniform non-dimensional vertical coordinate s,
+            aka sigma-coordinates. 0 <= s <= 1
+        ss1: float
+            stretched non-dimensional vertical coordinate s,
+            0 <= s <= 1
+        a1: float
+            parameter of the transformation
+        a2: float
+            parameter of the transformation
+        a3: float
+            parameter of the transformation
+        ss2: float
+             second stretched non-dimensional vertical coordinate s,
+             0 <= s <= 1 (only used for zco with ldbletanh = True)
+        a4: float
+            parameter of the transformation (only used for zco with ldbletanh = True)
+
+        Returns
+        -------
+        float
+
+        """
+
+        z = a1 + a2 * su + a3 * ss1 + a4 * ss2
+        return z
+
+    # -------------------------------------------------------------------------
+    def compute_e3(self, ds: Dataset) -> Dataset:
+        """
+        Grid cell thickness computed as discrete derivative
+        (central-difference) of levels' depth
+
+        Parameters
+        ----------
+        ds: Dataset
+
+        Returns
+        -------
+        Dataset
+        """
+        for k in range(self._jpk - 1):
+            ds["e3T"][k, :, :] = ds["z3W"][k + 1, :, :] - ds["z3W"][k, :, :]
+            ds["e3W"][k + 1, :, :] = ds["z3T"][k + 1, :, :] - ds["z3T"][k, :, :]
+        # Bottom:
+        k = -1
+        ds["e3T"][k, :, :] = 2.0 * (ds["z3T"][k, :, :] - ds["z3W"][k, :, :])
+        # Surface:
+        k = 0
+        ds["e3W"][k, :, :] = 2.0 * (ds["z3T"][k, :, :] - ds["z3W"][k, :, :])
```

Comment:
Same as before, I think you can use `[{"z": k}]` rather than `[k, :, :]`

Reply:
@oceandie Oops, looks like I didn't submit the previous comment. Anyways, I would index using dimension name rather than positional index, so we don't care about dimension order. So, anytime you do something like this `ds["e3T"][k, :, :] ` I would change it with `ds["e3T"][{"z": k}]`

Reply:
Thanks, very useful! ;)